### PR TITLE
Overhaul MELS to deal with causality, kicks, etc.

### PR DESCRIPTION
**Pull request by lukebarnard1**
The MELS can now deal with arbitrary sequences of transitions per user, where a transition is a change in membership. A transition can be joined, left, invite_reject, invite_withdrawal, invited, banned, unbanned or kicked.

Repeated segments (modulo 1 and 2), such as joined,left,joined,left,joined will be handled and will be rendered as " ... and 10 others joined and left 2 times and then joined". The repeated segments are assumed to be at the beginning of the sequence. This could be improved to handle arbitrary repeated sequences.

### CODE REVIEW

File: src/components/views/elements/MemberEventListSummary.js [hunk 1]
```diff
@@ -52,88 +52,122 @@ module.exports = React.createClass({
...
+            }
         }
+        return true;
+    },
 
-        let remaining = originalNumber - this.props.summaryLength;
-        if (remaining > 0) {
-            //  name1, name2, name3, and 100 others
-            return names + ', ' + lastName + ', and ' + remaining + ' others';
+    _renderCommaSeparatedList(items, disableAnd) {
```
Comment (richvdh) [inline src/components/views/elements/MemberEventListSummary.js]: rather than `disableAnd`, it might be preferable to have a `itemLimit` param, and move the 'and x others' logic in here - this making it reusable.

```
const remaining = itemLimit === undefined ? 0 : Math.max(items.length - itemLimit, 0);
if (items.length === 0) {
    return "";
} else if (items.length === 1) {
    return items[0];
} else if (remaining) {
    items = items.slice(0, itemLimit);
    const other = " other" + (remaining > 1 ? "s" : "");
    return items.join(', ') + ' and ' + remaining + other;
} else {
    let last = items.pop();
    return items.join(', ') + ' and ' + last;
}
```

File: src/components/views/elements/MemberEventListSummary.js [hunk 2]
```diff
@@ -52,88 +52,122 @@ module.exports = React.createClass({
...
+            let repeats = 1;
+            let repeatExtra = 0;
+
+            let splitTransitions = transitions.split(',');
+            let describedTransitions = splitTransitions;
+            let plural = eventAggregates[transitions].length > 1;
+
+            for (let modulus = 1; modulus <= 2; modulus++) {
+                // Sequences that are repeating through modulus transitions will be truncated
+                if (this._isRepeatedSequence(describedTransitions, modulus)) {
```
Comment (richvdh) [inline src/components/views/elements/MemberEventListSummary.js]: I don't think that looking for a pattern repeating throughout the whole of the transitions is the right approach. My understanding of matthew's intentions was that we would look specifically for pairs of 'join, leave' or 'leave, join' and treat them as special transitions. YMMV though.

File: src/components/views/elements/MemberEventListSummary.js [hunk 3]
```diff
@@ -175,61 +235,54 @@ module.exports = React.createClass({
...
 
-        // Populate the join/leave event arrays with events that represent what happened
-        // overall to a user's membership. If no events are added to either array for a
-        // particular user, they will be considered a user that "joined and left".
-        let joinEvents = [];
-        let leaveEvents = [];
-        let joinedAndLeft = 0;
-        let senders = Object.keys(userEvents);
-        senders.forEach(
+        // A map of agregate type to arrays of display names. Each aggregate type
```
Comment (lukebarnard1) [inline src/components/views/elements/MemberEventListSummary.js]: I think it will suffice to order by the first event of each user?

File: src/components/views/elements/MemberEventListSummary.js (hunk 2 above)
Comment (lukebarnard1) [inline src/components/views/elements/MemberEventListSummary.js]: If you push the transitions "joined","left" together to form "joined_and_left" as a transition, say, then you could reduce the problem from modulo 2 to modulo 1. Then you _could_ just look for possibly multiple contiguous blocks of transitions during any part of the sequence (instead of just the beginning). The only advantage the current implementation has over that is that it works with transitions in general. So a repeated "joined","kicked","joined","kicked" would be compressed. Otherwise you'd have to special case a "joined" next to a "kicked" as well... 

But I'm guessing join/leave/join/leave is _so_ common that it warrants a special transition?

File: src/components/views/elements/MemberEventListSummary.js (hunk 1 above)
Comment (lukebarnard1) [inline src/components/views/elements/MemberEventListSummary.js]: sgtm

File: src/components/views/elements/MemberEventListSummary.js (hunk 3 above)
Comment (lukebarnard1) [inline src/components/views/elements/MemberEventListSummary.js]: Done.

File: src/components/views/elements/MemberEventListSummary.js [hunk 4]
```diff
@@ -175,61 +235,54 @@ module.exports = React.createClass({
...
-                    // Increment the number of users whose membership change was nil overall
-                    joinedAndLeft++;
+                let displayName = userEvents[userId][0].getContent().displayname || userId;
+
+                let seq = this._getTransitionSequence(userEvents[userId]);
+                if (!aggregate[seq]) {
+                    aggregate[seq] = [];
+                }
+
+                // Assumes display names are unique
```
Comment (lukebarnard1) [inline src/components/views/elements/MemberEventListSummary.js]: Actually, using `target.name` works

File: src/components/views/elements/MemberEventListSummary.js [hunk 5]
```diff
@@ -175,61 +271,73 @@ module.exports = React.createClass({
...
+            // $userId : []
         };
 
-        eventsToRender.forEach((e) => {
+        eventsToRender.forEach((e, index) => {
             const userId = e.getStateKey();
             // Initialise a user's events
             if (!userEvents[userId]) {
-                userEvents[userId] = {first: null, last: null};
+                userEvents[userId] = [];
```
Comment (richvdh) [inline src/components/views/elements/MemberEventListSummary.js]: Idea: push to `avatarMembers` here, instead of at line 327? I think that would have the same effect?

Then you can move the `aggregate` functionality out to another function without having to worry about populating `avatarMembers`.

File: src/components/views/elements/MemberEventListSummary.js [hunk 6]
```diff
@@ -175,61 +271,73 @@ module.exports = React.createClass({
...
+                    aggregate[seq] = [];
+                    aggregateIndices[seq] = -1;
+                }
+
+                if (aggregate[seq].indexOf(displayName) === -1) {
+                    aggregate[seq].push(displayName);
                 }
+
+                if (aggregateIndices[seq] === -1 || firstEvent.index < aggregateIndices[seq]) {
+                    aggregateIndices[seq] = firstEvent.index;
```
Comment (richvdh) [inline src/components/views/elements/MemberEventListSummary.js]: believe this is incorrect, per https://github.com/vector-im/riot-web/issues/2856#issuecomment-273232027.

File: src/components/views/elements/MemberEventListSummary.js [hunk 7]
```diff
@@ -175,61 +271,73 @@ module.exports = React.createClass({
...
+                let firstEvent = userEvents[userId][0];
+                let displayName = firstEvent.displayName;
+
+                let seq = this._getTransitionSequence(userEvents[userId]);
+                if (!aggregate[seq]) {
+                    aggregate[seq] = [];
+                    aggregateIndices[seq] = -1;
+                }
+
+                if (aggregate[seq].indexOf(displayName) === -1) {
```
Comment (richvdh) [inline src/components/views/elements/MemberEventListSummary.js]: I think if two different Matthews have the same aggregate sequence, then we ought to show it as "Matthew and one other joined", not just "Matthew joined". (IOW: you need to remove this condition).

File: src/components/views/elements/MemberEventListSummary.js [hunk 8]
```diff
@@ -40,100 +40,183 @@ module.exports = React.createClass({
...
+        //     transitionType: "joined_and_left"
+        //     repeats: 123
+        // }, ... ]
+        return res;
+    },
+
+    _renderSummary: function(eventAggregates, orderedTransitionSequences) {
+        let summaries = orderedTransitionSequences.map((transitions) => {
+            let nameList = this._renderNameList(eventAggregates[transitions]);
+            let plural = eventAggregates[transitions].length > 1;
```
Comment (richvdh) [inline src/components/views/elements/MemberEventListSummary.js]: plural what?

I think this would be clearer if you factored out `let userNames = eventAggregates[transitions]` and used it here and above.

File: src/components/views/elements/MemberEventListSummary.js [hunk 9]
```diff
@@ -40,100 +40,183 @@ module.exports = React.createClass({
...
 
-        // The joinEvents and leaveEvents are representative of the net movement
-        // per-user, and so it is possible that the total net movement is nil,
-        // whilst there are some events in the expanded list. If the total net
-        // movement is nil, then neither joinSummary nor leaveSummary will be
-        // truthy, so return null.
-        if (!joinSummary && !leaveSummary) {
+            let repeats = 1;
```
Comment (richvdh) [inline src/components/views/elements/MemberEventListSummary.js]: these look redundant?

File: src/components/views/elements/MemberEventListSummary.js [hunk 10]
```diff
@@ -40,100 +40,183 @@ module.exports = React.createClass({
...
-            return names + ' and ' + lastName;
+            let last = items.pop();
+            return items.join(', ') + ' and ' + last;
         }
     },
 
-    _renderSummary: function(joinEvents, leaveEvents) {
-        let joiners = this._renderNameList(joinEvents);
-        let leavers = this._renderNameList(leaveEvents);
+    _getDescriptionForTransition(t, plural, repeats) {
```
Comment (richvdh) [inline src/components/views/elements/MemberEventListSummary.js]: could do with some more comments explaining what these helpers do, and what their params are. For example, the fact that `plural` is a boolean flag meaning that more than one user was affected would be very useful knowledge here.

File: src/components/views/elements/MemberEventListSummary.js [hunk 11]
```diff
@@ -40,100 +40,183 @@ module.exports = React.createClass({
...
+        let beConjugated = plural ? "were" : "was";
```
Comment (richvdh) [inline src/components/views/elements/MemberEventListSummary.js]: this is going to be hilarious to i18nify.

File: test/components/views/elements/MemberEventListSummary-test.js [hunk 12]
```diff
@@ -0,0 +1,556 @@
...
+        const instance = ReactDOM.render(<MemberEventListSummary {...props} />, parentDiv);
+        const summary = ReactTestUtils.findRenderedDOMComponentWithClass(instance, "mx_MemberEventListSummary_summary");
+        const summaryText = summary.innerText;
+
+        expect(summaryText).toBe("user_1 and 1 other were unbanned, joined and left 2 times and were banned");
+
+        done();
+    });
+
+    it('handles multple users following the same sequence of memberships', function(done) {
```
Comment (lukebarnard1) [inline test/components/views/elements/MemberEventListSummary-test.js]: or perhaps non-duplicate tests..

File: src/components/views/elements/MemberEventListSummary.js [hunk 13]
```diff
@@ -40,100 +40,183 @@ module.exports = React.createClass({
...
         this.setState({
             expanded: !this.state.expanded,
         });
     },
 
-    _getEventSenderName: function(ev) {
-        if (!ev) {
-            return 'undefined';
+    _renderNameList: function(users) {
+        if (users.length === 0) {
```
Comment (lukebarnard1) [inline src/components/views/elements/MemberEventListSummary.js]: fair enough

File: src/components/views/elements/MemberEventListSummary.js (hunk 5 above)
Comment (lukebarnard1) [inline src/components/views/elements/MemberEventListSummary.js]: wfm!

File: src/components/views/elements/MemberEventListSummary.js (hunk 7 above)
Comment (lukebarnard1) [inline src/components/views/elements/MemberEventListSummary.js]: Yep, that works.

File: src/components/views/elements/MemberEventListSummary.js (hunk 6 above)
Comment (lukebarnard1) [inline src/components/views/elements/MemberEventListSummary.js]: I'm not sure why we should order by the last event vs. the first event, or rather I'm not sure which is better. There are also cases where neither ordering by the first nor the last actually produces something accurate:

A joins, B joins, B leaves, B joins, A leaves.
ordered by first event: A joined and left, B joined and left and joined
ordered by last event: B joined and left, A joined and left

File: test/components/views/elements/MemberEventListSummary-test.js [hunk 14]
```diff
@@ -0,0 +1,556 @@
...
+        return events.map((e) => {
+            return (
+                <div key={e.getId()} className="event_tile">
+                    Expanded membership
+                </div>
+            );
+        });
+    };
+
+    const generateMembershipEvent = (eventId, parameters) => {
```
Comment (lukebarnard1) [inline test/components/views/elements/MemberEventListSummary-test.js]: I'd have to modify it to include the ability to set previous content, though mocking out `MatrixEvent` seems wrong. I'll just use a real `MatrixEvent` and override `target`.

File: test/components/views/elements/MemberEventListSummary-test.js [hunk 15]
```diff
@@ -0,0 +1,556 @@
...
+            eventsForUsers = eventsForUsers.concat(generateEvents(events));
+        }
+        return eventsForUsers;
+    };
+
+    beforeEach(function() {
+        testUtils.beforeEach(this);
+        sandbox = testUtils.stubClient();
+        parentDiv = document.createElement('div');
+        document.body.appendChild(parentDiv);
```
Comment (lukebarnard1) [inline test/components/views/elements/MemberEventListSummary-test.js]: wfm

File: src/components/views/elements/MemberEventListSummary.js (hunk 6 above)
Comment (ara4n) [inline src/components/views/elements/MemberEventListSummary.js]: @lukebarnard1 why would it ever be "B joined and left"?  B always "joined and left and joined" surely?

Comment (ara4n) [inline src/components/views/elements/MemberEventListSummary.js]: let's discuss this on the issue, not the PR.

File: test/components/views/elements/MemberEventListSummary-test.js [hunk 16]
```diff
@@ -20,42 +22,41 @@ describe.only('MemberEventListSummary', function() {
...
-            getSender: () => {
-                return senderId || userId;
-            },
-            getStateKey: () => {
-                return userId;
-            },
-        };
+        });
+        // Override random event ID
+        e.event.event_id = eventId;
```
Comment (richvdh) [inline test/components/views/elements/MemberEventListSummary-test.js]: why is this necessary, ooi?

File: test/components/views/elements/MemberEventListSummary-test.js [hunk 17]
```diff
@@ -4,12 +4,14 @@ const ReactDOM = require("react-dom");
 const ReactTestUtils = require('react-addons-test-utils');
 const sdk = require('matrix-react-sdk');
 const MemberEventListSummary = sdk.getComponent('views.elements.MemberEventListSummary');
+const jssdk = require('matrix-js-sdk');
+const MatrixEvent = jssdk.MatrixEvent;
 
 const testUtils = require('../../../test-utils');
 describe.only('MemberEventListSummary', function() {
```
Comment (richvdh) [inline test/components/views/elements/MemberEventListSummary-test.js]: still .only ? whyyyyyyyyyyyy?

File: src/components/views/elements/MemberEventListSummary.js [hunk 18]
```diff
@@ -65,67 +65,73 @@ module.exports = React.createClass({
...
+     */
+    _renderNameList: function(users) {
+        return this._renderCommaSeparatedList(users, this.props.summaryLength);
+    },
+
+    /**
+     * Canonicalise an array of transitions into an array of transitions and how many times
+     * they are repeated consecutively.
+     *
+     * An array of 123 "joined_and_left" transitions, would result in:
```
Comment (richvdh) [inline src/components/views/elements/MemberEventListSummary.js]: lies

File: src/components/views/elements/MemberEventListSummary.js [hunk 19]
```diff
@@ -65,67 +65,73 @@ module.exports = React.createClass({
...
+     * ```
+     * [{
+     *   transitionType: "joined_and_left"
+     *   repeats: 123
+     * }, ... ]
+     * ```
+     * @param {string[]} transitions the array of transitions to transform.
+     * @returns {object[]} an array of truncated transitions.
```
Comment (richvdh) [inline src/components/views/elements/MemberEventListSummary.js]: lies

File: test/components/views/elements/MemberEventListSummary-test.js (hunk 16 above)
Comment (lukebarnard1) [inline test/components/views/elements/MemberEventListSummary-test.js]: Because when the tests for expanded events are done, the shallow equality tests require a list of React instances (https://github.com/matrix-org/matrix-react-sdk/blob/bd106968ceafbad30766958008a6c840acc6e1aa/test/components/views/elements/MemberEventListSummary-test.js#L134) 

And the keys of these must match those generated in `generateTiles`, so they need to be constant, not random.

File: src/components/views/elements/MemberEventListSummary.js [hunk 20]
```diff
@@ -227,6 +267,18 @@ module.exports = React.createClass({
         );
     },
 
+    _getTransitionSequence: function(events) {
+        return events.map(this._getTransition);
+    },
+
+    /**
+     * Enumerate a given membership event, `e`, where `getContent().membership` has
```
Comment (richvdh) [inline src/components/views/elements/MemberEventListSummary.js]: I don't understand this description. How can you enumerate one thing?

You keep using that word. I do not think it means what you think it means.

File: src/components/views/elements/MemberEventListSummary.js [hunk 21]
```diff
@@ -159,6 +165,20 @@ module.exports = React.createClass({
...
+     * ```
+     * [{
+     *   transitionType: "joined_and_left"
+     *   repeats: 123
+     * }, ... ]
+     * ```
+     * @param {string[]} transitions the array of transitions to transform.
+     * @returns {object[]} an array of truncated transitions.
+     */
     _getTruncatedTransitions: function(transitions) {
```
Comment (richvdh) [inline src/components/views/elements/MemberEventListSummary.js]: I think this is a poor name for this. It implies that you're dropping transitions off the list, which isn't the case.

How about something like `_coalesceRepeatedTransitions` or sth?

File: src/components/views/elements/MemberEventListSummary.js [hunk 22]
```diff
@@ -65,67 +65,73 @@ module.exports = React.createClass({
...
+     * more items in `users` than `this.props.summaryLength`, which is the number of names
+     * included before "and [n] others".
+     */
+    _renderNameList: function(users) {
+        return this._renderCommaSeparatedList(users, this.props.summaryLength);
+    },
+
+    /**
+     * Canonicalise an array of transitions into an array of transitions and how many times
+     * they are repeated consecutively.
```
Comment (richvdh) [inline src/components/views/elements/MemberEventListSummary.js]: lies

File: src/components/views/elements/MemberEventListSummary.js (hunk 21 above)
Comment (richvdh) [inline src/components/views/elements/MemberEventListSummary.js]: "Truncated" appears in other places in this file too (including in the jsdoc). Remember to update them.

File: test/components/views/elements/MemberEventListSummary-test.js (hunk 16 above)
Comment (richvdh) [inline test/components/views/elements/MemberEventListSummary-test.js]: ok. Could you add this as a comment?

File: test/components/views/elements/MemberEventListSummary-test.js (hunk 17 above)
Comment (lukebarnard1) [inline test/components/views/elements/MemberEventListSummary-test.js]: oops, I'll get rid - it was just to isolate the test so that the others wouldn't run

File: src/components/views/elements/MemberEventListSummary.js (hunk 21 above)
Comment (lukebarnard1) [inline src/components/views/elements/MemberEventListSummary.js]: Agreed that truncated is poor and that to coalesce the events sounds better.

File: src/components/views/elements/MemberEventListSummary.js (hunk 20 above)
Comment (lukebarnard1) [inline src/components/views/elements/MemberEventListSummary.js]: I think I meant 'label', I'll use that instead

File: test/components/views/elements/MemberEventListSummary-test.js (hunk 16 above)
Comment (lukebarnard1) [inline test/components/views/elements/MemberEventListSummary-test.js]: sure

**Comment by richvdh** [comment]
Ok. You can also set KARMA_TEST_FILE in the environment to limit the 
runner to one file without having to remember to fix the source.

**Comment by lukebarnard1** [comment]
image
